AppCenter/AppCenterTests/MSAbstractLogTests.m

```diff
@@ -0,0 +1,168 @@
+
+  // When
+  NSData *serializedLog = [NSKeyedArchiver archivedDataWithRootObject:self.sut];
+  id actual = [NSKeyedUnarchiver unarchiveObjectWithData:serializedLog];
+
+  // Then
+  assertThat(actual, notNilValue());
+  assertThat(actual, instanceOf([MSAbstractLog class]));
+
+  MSAbstractLog *actualLog = actual;
+  assertThat(actualLog.type, equalTo(type));
+  assertThat(actualLog.timestamp, equalTo(timestamp));
+  assertThat(actualLog.sid, equalTo(sid));
+  assertThat(actualLog.distributionGroupId, equalTo(distributionGroupId));
+  assertThat(actualLog.device, equalTo(device));
+}
+
+- (void)testIsValid {
+
+  // If
+  id device = OCMClassMock([MSDevice class]);
+  OCMStub([device isValid]).andReturn(YES);
+  self.sut.type = @"fake";
+  self.sut.timestamp = [NSDate dateWithTimeIntervalSince1970:42];
+  self.sut.device = device;
+
+  // Then
+  XCTAssertTrue([self.sut isValid]);
+
+  // When
+  self.sut.type = nil;
+  self.sut.timestamp = [NSDate dateWithTimeIntervalSince1970:42];
+  self.sut.device = device;
+
+  // Then
+  XCTAssertFalse([self.sut isValid]);
+
+  // When
+  self.sut.type = @"fake";
+  self.sut.timestamp = nil;
+  self.sut.device = device;
+
+  // Then
+  XCTAssertFalse([self.sut isValid]);
+
+  // When
+  self.sut.type = @"fake";
+  self.sut.timestamp = [NSDate dateWithTimeIntervalSince1970:42];
+  self.sut.device = nil;
+
+  // Then
+  XCTAssertFalse([self.sut isValid]);
+}
+
+- (void)testIsEqual {
+
+  // If
+  NSString *type = @"fake";
+  NSDate *timestamp = [NSDate dateWithTimeIntervalSince1970:0];
+  NSString *sid = @"FAKE-SESSION-ID";
+  NSString *distributionGroupId = @"FAKE-GROUP-ID";
+  MSDevice *device = [MSDevice new];
+
+  self.sut.type = type;
+  self.sut.timestamp = timestamp;
+  self.sut.sid = sid;
+  self.sut.distributionGroupId = distributionGroupId;
+  self.sut.device = device;
+
+  // When
+  NSData *serializedEvent = [NSKeyedArchiver archivedDataWithRootObject:self.sut];
+  id actual = [NSKeyedUnarchiver unarchiveObjectWithData:serializedEvent];
+  MSAbstractLog *actualLog = actual;
+
+  // Then
+  XCTAssertTrue([self.sut isEqual:actualLog]);
+
+  // When
+  self.sut.type = @"new-fake";
+
+  // Then
+  XCTAssertFalse([self.sut isEqual:actualLog]);
+
+  // When
+  self.sut.type = @"fake";
+  self.sut.distributionGroupId = @"FAKE-NEW-GROUP-ID";
+
+  // Then
+  XCTAssertFalse([self.sut isEqual:actualLog]);
+}
+
+- (void)testSerializingToJsonWorks {
+
+  // If
+  self.sut.type = @"fake";
+  self.sut.timestamp = [NSDate dateWithTimeIntervalSince1970:0];
+  self.sut.sid = @"FAKE-SESSION-ID";
+  self.sut.distributionGroupId = @"FAKE-GROUP-ID";
+  self.sut.device = [MSDevice new];
+
+  // When
+  NSMutableDictionary *actual = [self.sut serializeLogWithPrettyPrinting:false];
+
+  // Then
+  assertThat(actual, notNilValue());
+  assertThat(actual, equalTo(@"{\"type\":\"fake\",\"sid\":\"FAKE-SESSION-ID\",\"device\":{},\"distributionGroupId\":\"FAKE-GROUP-ID\",\"timestamp\":\"1970-01-01T00:00:00.000Z\"}"));
```

Comment:
This causes build failure. String comparison doesn't make sense here since the string is JSON payload and order of key in the payload is not guaranteed.

Reply:
You are right. I've changed my code to parse generated JSON so keys ordering doesn't matter now. Thank you.